src/nomad_simulations/schema_packages/data_types.py

```diff
@@ -0,0 +1,260 @@
+        pass
+
+    def _validate_bounds(self, normalized_value: Any) -> None:
+        """Common bounds validation logic."""
+        flat_values = list(_flatten_values(normalized_value))
+        valid_values = self._filter_values_for_validation(flat_values)
+
+        if valid_values:
+            invalid_values = [
+                v
+                for v in valid_values
+                if not _check_bounds(
+                    v,
+                    self._min_value,
+                    self._max_value,
+                    self._min_inclusive,
+                    self._max_inclusive,
+                )
+            ]
+
+            if invalid_values:
+                min_val = min(valid_values)
+                max_val = max(valid_values)
+                bounds = _get_bounds_str(
+                    self._min_value,
+                    self._max_value,
+                    self._min_inclusive,
+                    self._max_inclusive,
+                )
+                raise ValueError(
+                    f'{self._error_message_prefix} must be in {bounds}, got range [{min_val}, {max_val}]'
+                )
+
+
+class BoundedInt(BoundedDataType, ExactNumber):
+    """
+    32-bit integer data type with configurable bounds (bound value and inclusion).
+    Bounds are by default +/- infinity (`float('inf')`, `float('-inf')` respectively).
+
+    Precision: 32-bit signed integer (range: -2,147,483,648 to 2,147,483,647)
+    Convertible from: Python int, np.int32, np.int16, np.int8
+
+    Examples:
+        BoundedInt().min(1)                    # >= 1 (positive integers)
+        BoundedInt().min(0)                    # >= 0 (non-negative integers)
+        BoundedInt().min(1, inclusive=False)   # > 1
+        BoundedInt().max(10)                   # <= 10
+        BoundedInt().min(1).max(10)            # 1 <= value <= 10
+    """
+
+    __slots__ = ('_min_value', '_max_value', '_min_inclusive', '_max_inclusive')
+    _error_message_prefix = 'All values'
+
+    def __init__(self, *, dtype: type[int | np.int32] = np.int32):
+        if isinstance(dtype, np.dtype):
+            dtype = dtype.type
+
+        if dtype not in (int, np.int32):
+            raise ValueError(
+                f'Invalid dtype for {self.__class__.__name__}. Only int and np.int32 are supported.'
+            )
+
+        ExactNumber.__init__(self, dtype)
+        BoundedDataType.__init__(self)
+
+    def convertible_from(self, other: Any) -> bool:
+        """Check if this data type can convert from another type."""
+        return other in {int, np.int32, np.int16, np.int8}
+
+    def _filter_values_for_validation(self, flat_values: list[Any]) -> list[Any]:
+        """No filtering needed for integers - all values are valid for bounds checking."""
+        return flat_values
+
+    def normalize(self, value: Any, **kwargs: Any) -> Any:
+        normalized_value = ExactNumber.normalize(self, value, **kwargs)
```

Comment:
This `kwargs` contains the parent section that holds the data, in case it becomes dynamic somehow and relies on other quantities, you can get section from it can get the information accordingly, there are example usages in the codebase.

Reply:
That's good to know, will keep it in mind for the follow-up to this PR.